[sensebox] binding initial contribution

hakan42 opened the pull request:
Signed-off-by: Hakan Tandogan <hakan@tandogan.com>

hakan42 commented:
Updated to 2.1.0-SNAPSHOT and squashed commits

hakan42 commented:
Trigger build for market...

hakan42 commented:
rebuild

kaikreuzer commented:
Jenkins error is my fault...

hakan42 commented:
Ah, I was hoping that rebasing against master would lead to a successful build (for the market). Anyway, thanks for working on the feed build-breaker 😄

kaikreuzer commented:
let's give it a new try without static analysis checks ;-)

kaikreuzer commented:
This time it is your fault :-P

hakan42 commented:
No way, just testing whether your checks really find everything 😁 

But seriously, I guess I have to check my build.properties. No idea why my local builds are successful though.

[ERROR] Failed to execute goal org.eclipse.tycho:tycho-packaging-plugin:1.0.0:package-plugin (default-package-plugin) on project org.openhab.binding.sensebox: /scratch/jenkins/workspace/PR-openHAB2-Addons/addons/binding/org.openhab.binding.sensebox/build.properties: bin.includes value(s) [lib/] do not match any files.

hakan42 commented:
The build is as green as it gets 😄

martinvw commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,374 @@
+        State result = UnDefType.UNDEF;
+
+        if (data != null) {
+            if (data.getLastMeasurement() != null) {
```
You could consider grouping the conditions instead of nesting them

hakan42 replied:
Also a question of coding style, I find multiple nested ifs help readibility, and actual debugging as every clause is on its own line.

martinvw commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/internal/ExpiringCache.java:
```diff
@@ -0,0 +1,68 @@
+ * called only if the answer from the last calculation is not valid anymore, i.e. if the key is different or the last
+ * result is expired.
+ *
+ * @author Simon Kaufmann - Initial contribution and API.
```
Also-by?

hakan42 replied:
True, forgot that. Added.

hakan42 replied:
This class has its own commit, with the Also-By now.

martinvw replied:
👍

hakan42 commented on addons/binding/org.openhab.binding.sensebox/.idea/compiler.xml:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Mistake in my commit, I will remove the .idea directory.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<binding:binding id="sensebox"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Done.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/ESH-INF/config/config-description.xml:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Done.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,195 @@
+        <state readOnly="true"/>
+    </channel-type>
+
+    <!--
```
Comments for the developer. Especially at the measurements, I looked at the response what units the data is given in so I don't forget that information.

martinvw replied:
All of it seems to be already duplicated in the xml elements below are they still relevant?

hakan42 commented on addons/binding/org.openhab.binding.sensebox/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,27 @@
+Bundle-SymbolicName: org.openhab.binding.sensebox;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.1.0.qualifier
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
```
Done.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/OSGI-INF/SenseBoxHandlerFactory.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2014-2016 by the respective copyright holders.
```
Done.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/README.md:
```diff
@@ -0,0 +1,114 @@
+In the table is shown more detailed information about each Channel type.
+The binding introduces the following channels:
+
+| Channel ID | Channel Description | Supported item type | Advanced |
```
Done.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/org.openhab.binding.sensebox.iml:
```diff
@@ -0,0 +1,121 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Yes, adding *.iml (and maybe .idea) to the .gitignore at the root would be fine. Until then, I'll remove this file from my commit, too.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/pom.xml:
```diff
@@ -0,0 +1,23 @@
+	</parent>
+
+	<properties>
+		<bundle.symbolicName>org.openhab.binding.sensebox</bundle.symbolicName>
```
No idea, I believe I copied that from some other binding and did not think a lot about it 😄 . Removed.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,374 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (thing.getStatus().equals(ThingStatus.ONLINE)) {
```
Done.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,374 @@
+            }
+        };
+
+        refreshJob = scheduler.scheduleAtFixedRate(runnable, 0, thingConfiguration.getRefreshInterval(),
```
Done.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,374 @@
+        try {
+            data = CACHE.get(thingConfiguration.getSenseBoxId());
+        } catch (IOException e) {
+            logger.debug("Exception occurred during fetching data: {}", e.getMessage(), e);
```
I changed the logger output to error, but would like to keep the item online. Honestly, I have to thing a bit how to handle the online/offline state changes 😄

martinvw replied:
Will you do this before merging this pr?

hakan42 replied:
Just running a fresh compile, and about to push.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,374 @@
+            new ExpiringCache.LoadAction<String, SenseBoxData>() {
+                @Override
+                public SenseBoxData load(String senseBoxId) throws IOException {
+                    SenseBoxData result = reallyFetchDataFromServer(senseBoxId);
```
True :)

But in such cases, I actually prefer to have the local variable, as it helps in the debugger, and for productive code, the compiler optimizes it anyway 😄

hakan42 commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,374 @@
+                }
+            }
+
+            for (SenseBoxSensor sensor : parsedData.getSensors()) {
```
Actually, I did. But sometimes, even eclipse throws Exceptions while formatting, so this file might have slipped through the cracks.

martinvw replied:
I saw it in other commits, maybe I'm wrong :-)

hakan42 commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,374 @@
+                } else if ("°C".equals(sensor.getUnit())) {
+                    parsedData.setTemperature(sensor);
+                } else {
+                    logger.debug("    Sensor: {}", sensor);
```
May be, but in that case, it helps immensely tracing API connection problems without creating costly toString() methods on the value holder objects. I would prefer to keep this as it is.

hakan42 commented:
* Addressed review comments
* rebased against current master

martinvw commented on addons/binding/org.openhab.binding.sensebox/ESH-INF/thing/box.xml:
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
👍

hakan42 commented on addons/binding/org.openhab.binding.sensebox/README.md:
```diff
@@ -0,0 +1,114 @@
+
+## Channel configuration
+
+...
```
Honestly, I don't know what to write in this section. I am notoriously bad at writing documentation, maybe we can remove the "Channel configuration" until I can come up with something worthwhile?

martinvw replied:
👍

kaikreuzer replied:
Well, your channels do not have any configuration parameters, so there is no need for this section anyhow :-)

martinvw commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,372 @@
+        try {
+            data = CACHE.get(thingConfiguration.getSenseBoxId());
+        } catch (IOException e) {
+            logger.error("Exception occurred during fetching data: {}", e.getMessage(), e);
```
Some additional minor remarks about this, I believe error should not be used for this, see http://docs.openhab.org/developers/development/guidelines.html#e-logging

Maybe you should not even catch the exception here, but have this body in the `run` method;

```java
try {
    fetchData();
    ... handle the data ...
} catch (IOException e) {
    logger.warn("Exception occurred during fetching data: {}", e.getMessage(), e); 
} catch (Exception e) {
    logger.debug("Exception occurred during execution: {}", e.getMessage(), e);
}

hakan42 replied:
My intention was to keep a somewhat parseable data object around even if the connection fails. Thinking more about this, I believe it is better to start handling data only if fetching and parsing is successful.

Therefore, I implemented your suggestion.

martinvw replied:
Can it be that you did not yet push it?

hakan42 replied:
Aaaaah.... Oops 😊 

Yes, it might be that I might have forgotten to run that 'git push --force' 😄 

Now I did it, thank you for your patience with this PR, I hope we whipped it into mergeable state now.

hakan42 commented:
Adapted change to the fetchData() method and force-pushed once more.

martinvw commented:
@hakan42 thanks for your updates, they look perfect now :-)

However some guidelines changed (see also https://github.com/openhab/openhab2-addons/pull/2179) can you please add the `about.html` to the `build.properties. Thanks!

Then I'll click approve and its up to kai

hakan42 commented:
I added the about.html already with the last push yesterday 😄

martinvw commented:
I was just copy-pasting the message didn't check it :-D

Thanks 👍

martinvw commented:
Triggering a rebuild

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/README.md:
````diff
@@ -0,0 +1,114 @@
+}
+```
+
+## Reporting issues
````
Please remove this section - the way to report issues should be the same for all bindings, so there should be no need to provide such sections for every single binding.

hakan42 replied:
Done.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/SenseBoxBindingConstants.java:
```diff
@@ -0,0 +1,53 @@
+    public static final String SENSEMAP_IMAGE_URL_BASE = "https://opensensemap.org/userimages";
+    public static final String SENSEMAP_MAP_URL_BASE = "https://opensensemap.org";
+
+    // Minimum timeslice between API requests
```
in seconds/minutes/...?

hakan42 replied:
Amended documentation.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+ *
+ * @author Hakan Tandogan - Initial contribution
+ */
+
```
remove empty line

hakan42 replied:
Done.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/config/SenseBoxConfiguration.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public class SenseBoxConfiguration {
+
+    private String thingUid;
```
The thingUID is not a config parameter of the Thing, so it is rather strange to have that added here - it really doesn't belong in here.

hakan42 replied:
Old code from my first experiments with the OH2 API 😄 

removed.

hakan42 replied:
Right, leftover from earlier code. Removed.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+
+        thingConfiguration = getConfigAs(SenseBoxConfiguration.class);
+        String thingUid = getThing().getUID().toString();
+        thingConfiguration.setThingUid(thingUid);
```
should be removed - you can at any time get the thingUID from the Thing.

hakan42 replied:
Done.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+
+        if (StringUtils.trimToNull(thingConfiguration.getSenseBoxId()) == null) {
+            offlineReason = "senseBox ID is mandatory and must be configured";
+            logger.error("{}, disabling thing '{}'", offlineReason, thingUid);
```
Check our logging guidelines (E.5): http://docs.openhab.org/developers/development/guidelines.html#e-logging
You should not do logging, but set the ThingStatus instead.

hakan42 replied:
Logging removed, the ThingStatus is being set to OFFLINE at the end of the  initialize() method.

kaikreuzer replied:
It should only be set to OFFLINE, if you know that it is OFFLINE (i.e. you cannot reach it). Otherwise better set it to UNKNOWN.

hakan42 replied:
o.k., set to UNKNOWN instead of OFFLINE.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+        }
+
+        if (thingConfiguration.getRefreshInterval() < MINIMUM_UPDATE_INTERVAL) {
+            logger.info("Refresh interval is much too small, setting to default of {} seconds",
```
So why did you define min=1 in https://github.com/openhab/openhab2-addons/pull/1543/files#diff-a8ca44d8ffc786e0dc407808a119825aR15?

hakan42 replied:
Forgot that part. Amended the "1" in the config-description.xml to "300" as well.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+
+        if (validConfig) {
+            logger.debug("Thing Configuration: {}", thingConfiguration.toString());
+            updateStatus(ThingStatus.ONLINE);
```
Really? Are you sure you can reach it?

hakan42 replied:
No, but to be able to check for connectivity, I have to put the Thing into ONLINE mode, right? I would like to avoid expensive HTTP operations during initialization.

kaikreuzer replied:
No, see my comments further below in handleCommand :-)

hakan42 replied:
So, if I understand http://docs.openhab.org/concepts/things.html correctly, things should be UNKNOWN at the end of initialize(), and the communication attemps take care of switching it to either ONLINE or OFFLINE, right?

kaikreuzer replied:
Yes. Alternatively, you can skip UNKNOWN if you know that you are "quickly" setting either ONLINE or OFFLINE soon after initialisation.

hakan42 replied:
o.k., I set it to UNKNOWN on initialize, and leave it to startAutomaticRefresh() to switch the item to ONLINE or OFFLINE. better to have only one way in the code flow.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+                    data = fetchData();
+                    publishDataForChannel(channelUID);
+                } catch (IOException e) {
+                    logger.warn("Exception occurred during fetching data: {}", e.getMessage(), e);
```
I would assume that IOException can happen due to network issues. If I am right, the same as above applies: Do not do error logging, but set the ThingStatus.

hakan42 replied:
Done.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+                } catch (IOException e) {
+                    logger.warn("Exception occurred during fetching data: {}", e.getMessage(), e);
+                } catch (Exception e) {
+                    logger.debug("Exception occurred during execution: {}", e.getMessage(), e);
```
In which situations can this happen? Might you also want to set the ThingStatus to OFFLINE here?

hakan42 replied:
Leftover from an earlier code where the caching layer could throw plain Exceptions. Removed.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (ThingStatus.ONLINE == thing.getStatus()) {
```
You should not do such checks - if it is OFFLINE, you might be able to reach it by now and you just haven't noticed yet. So better TRY to reach it and stay OFFLINE, if you fail.

hakan42 replied:
Done.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+
+    @Override
+    public void dispose() {
+        refreshJob.cancel(true);
```
refreshJob can be null here

hakan42 replied:
Put check in here.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+        Runnable runnable = new Runnable() {
+            @Override
+            public void run() {
+                logger.info("Refreshing data for box {}, scheduled after {} seconds...",
```
reduce to debug

hakan42 replied:
Done.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+                try {
+                    data = fetchData();
+
+                    publishDataForChannel(new ChannelUID(getThing().getUID(), CHANNEL_NAME));
```
you should avoid so many object instantiations. Why not simply call
```
publishDataForChannel(CHANNEL_NAME));
```
?

hakan42 replied:
Did not know that this method was available 😄 

Done.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+                    publishDataForChannel(new ChannelUID(getThing().getUID(), CHANNEL_TEMPERATURE_LR));
+
+                } catch (IOException e) {
+                    logger.warn("Exception occurred during fetching data: {}", e.getMessage(), e);
```
same as above

hakan42 replied:
reduced log level and switched Thing status to OFFLINE.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+                } catch (IOException e) {
+                    logger.warn("Exception occurred during fetching data: {}", e.getMessage(), e);
+                } catch (Exception e) {
+                    logger.debug("Exception occurred during execution: {}", e.getMessage(), e);
```
same as above

hakan42 replied:
Done.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+
+        try {
+            URL url = new URL(query);
+            URLConnection connection = url.openConnection();
```
Is this connection ever closed again? Better use a try-with-resources.

hakan42 replied:
try-with-resources would require the URLConnection to be Autoclosable, right?  Do we have an example somewhere where this pattern is used on HTTP connections?

In any case, a close() on the InputStream is added, and this should close the connection too if I read the javadoc right.

kaikreuzer replied:
Well, yes, my comment should have been one line below (on the getInputStream()).
You current solution is not valid as IOUtils.toString() can throw exceptions in which case you will never reach your "close" - that's why try-with-resources is the better way.

Also note that `toString(InputStream input)` [is deprecated](https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/IOUtils.html#toString(java.io.InputStream)[https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/IOUtils.html#toString(java.io.InputStream)]) - you should also pass a suitable encoding.

hakan42 replied:
Done, hopefully right this time :)

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/handler/SenseBoxHandler.java:
```diff
@@ -0,0 +1,373 @@
+            URLConnection connection = url.openConnection();
+            String body = IOUtils.toString(connection.getInputStream());
+
+            logger.debug("Fetched Data: {}", body);
```
Might that not be a lot of logging and thus better be trace level?

hakan42 replied:
True 😄    Done.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/src/main/java/org/openhab/binding/sensebox/internal/ExpiringCache.java:
```diff
@@ -0,0 +1,68 @@
+ *            <Q>the type of the key
+ * @param <V> the type of the value
+ */
+public class ExpiringCache<Q, V> {
```
If this is something useful, would it make sense to put it into the ESH core bundles as it is not a all binding related? Where exactly did you copy it from?

hakan42 replied:
@maggu2810 implemented this in the yahooweather binding over at ESH core: 

https://github.com/eclipse/smarthome/pull/2809

I believe a more generic caching layer would be useful.

sjka replied:
Yes, indeed. I remember I was trying to use guava for the same purpose, but we'd need a newer version of it and then things suddenly become ugly. I'd agree to put it somewhere into the core, so we can reuse it in different places. Let's rather discuss it here https://github.com/eclipse/smarthome/issues/3302.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/ESH-INF/thing/box.xml:
```diff
@@ -0,0 +1,25 @@
+		<description>Sensor Box</description>
+
+		<channel-groups>
+			<channel-group typeId="descriptors" id="descriptors"/>
```
Actually, the location, the imageUrl and can and do change. For the first, it is possible to have mobile senseBoxes, for example on your car. For the second, this changes as whenever you upload a new image for your installation.

I will try to convert the rest to properties.

hakan42 replied:
If I am not mistaken, I cannot use "properties" in the Paper UI Control tab to visualize  data. Neither can I bind them to Items to use in rules or sitemaps. Therefore, I would like to keep the "descriptors" as channels. @kaikreuzer ?

kaikreuzer replied:
> Neither can I bind them to Items to use in rules or sitemaps.

Well, if they are static, there really isn't much sense in doing so. And for showing static Thing properties on a Text widget in UIs, I would rather prefer to enhance the sitemaps to be able to reference a thing property for a text widget to display.

So maybe let's look at the channels/properties individually:
- location: Ok, if it can change, let's keep it as a channel.
- imageUrl: If you want to use images, I would actually suggest to use an ImageItem for it
- name + exposure: I would keep that as properties really
- apiUrl + mapUrl: Is this in any way relevant for UIs at all? And if, wouldn't it be good enough to show it on the Thing as properties of it?

hakan42 replied:
Changes image Url, map Url, name  and exposure to properties, leaving only the channel as a channel.

hakan42 commented on addons/binding/org.openhab.binding.sensebox/ESH-INF/thing/box.xml:
```diff
@@ -0,0 +1,25 @@
+			<channel-group typeId="lastReported" id="lastReported"/>
+		</channel-groups>
+
+		<properties>
```
Done.

hakan42 commented:
@kaikreuzer , I have addressed your points, could you have another look please?

@martinvw , now that pullapprove is active, is it enough if I force-push a squashed commit at the end of the review? I missed signing my fixes for the review comments...

hakan42 commented:
@kaikreuzer , I believe I addressed all your points about channels and properties. Could you please give it another quick look?

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,210 @@
+		<state readOnly="true" pattern="%2$s°N,%3$s°W, %1$s m"/>
+	</channel-type>
+
+	<channel-type id="imageUrl" advanced="true">
```
As you removed the channel, you should also remove the type.

kaikreuzer commented on addons/binding/org.openhab.binding.sensebox/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,210 @@
+		<state readOnly="true"/>
+	</channel-type>
+
+	<channel-type id="mapUrl" advanced="true">
```
As you removed the channel, you should also remove the type.

hakan42 commented:
* superfluous channel types removed
* squashed into two commits, one with my code alone, the second for the ExpiringCache initially written by @skja ("Also-by" added)

kaikreuzer commented:
@hakan42 So as it is now in the distro, you might want to remove it from the marketplace again.

hakan42 commented:
Thank you, I removed the entry just now.
